src/peer.ts

```diff
@@ -1,15 +1,194 @@
+            if (senderIp !== undefined && message.body !== undefined) {
+                this.sender.removeSentMessage(message.body);
+            }
+        });
+
+        // Handle join messages
+        this.registerReceiveHandlerImpl(MessageType.JOIN, async (message: Message, senderIp: string) => {
+            // Check if node already has an id, if so do not proceed with join request
+            if (message.originalSenderGuid === Guid.EMPTY) {
+                const newPeerId: string = Guid.create().toString();
+
+                // Add the new peer to our registry
+                this.routingTable.addPeer(newPeerId, senderIp);
+
+                // Send response
+                this.sendMessage(
+                    MessageType.JOIN_RESPONSE,
+                    newPeerId,
+                    JSON.stringify({guid: newPeerId, routingTable: Array.from(this.routingTable.peers)}),
+                );
+
+                // Let other peers know about the newly joined peer
+                this.sendBroadcast(
+                    MessageType.NEW_PEER,
+                    JSON.stringify({guid: newPeerId, sender: senderIp}),
+                );
+            }
+        });
+
+        // Handle join response messages
+        this.registerReceiveHandlerImpl(MessageType.JOIN_RESPONSE, async (message: Message, senderIp: string) => {
+            if (message.body !== undefined
+                && this.GUID === Guid.EMPTY
+                && senderIp !== undefined
+                && message.originalSenderGuid !== undefined) {
+                const body = JSON.parse(message.body);
+                this.GUID = body.guid;
+                this.routingTable.addPeer(message.originalSenderGuid, senderIp);
+                this.routingTable.mergeRoutingTables(new Map(body.routingTable));
+            }
+        });
+
+        // Handle new peer messages
+        this.registerReceiveHandlerImpl(MessageType.NEW_PEER, async (message: Message, senderIp: string) => {
+            if (senderIp !== undefined && message.body !== undefined) {
+                // Add the new peer to our registry
+                const body = JSON.parse(message.body);
+                this.routingTable.addPeer(body.guid, body.sender);
+            }
+        });
+
+        // Handle leave messages
+        this.registerReceiveHandlerImpl(MessageType.LEAVE, async (message: Message, senderIp: string) => {
+            if (message !== undefined && senderIp !== undefined) {
+                // Remove the new peer from our registry
+                this.routingTable.removePeer(message.originalSenderGuid);
+            }
+        });
+    }
 
-    constructor() {
-        this.sender = new Sender();
-        this.receiver = new Receiver();
-        this.ipRegistry = new IpRegistry();
+    /**
+     * Try to join the network. Send a join request to every given peer.
+     */
+    private checkInitialPeers(peers: string[]): void {
+        peers.forEach((peer) => {
```

Comment:
Use the `for of` loop instead. As the `foreach` creates a new closure that might be unwanted. 
The `for of` loop is also more efficient for the typescript compiler.